connector-runtime/connector-runtime-core/src/main/java/io/camunda/connector/runtime/core/inbound/DefaultInboundConnectorContextFactory.java

```diff
@@ -0,0 +1,87 @@
+import io.camunda.connector.api.inbound.InboundConnectorExecutable;
+import io.camunda.connector.api.inbound.InboundIntermediateConnectorContext;
+import io.camunda.connector.api.validation.ValidationProvider;
+import io.camunda.connector.runtime.core.inbound.correlation.InboundCorrelationHandler;
+import io.camunda.connector.runtime.core.secret.SecretProviderAggregator;
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.util.function.Consumer;
+
+public class DefaultInboundConnectorContextFactory implements InboundConnectorContextFactory {
+  private final ObjectMapper objectMapper;
+  private final InboundCorrelationHandler correlationHandler;
+  private final SecretProviderAggregator secretProviderAggregator;
+  private final ValidationProvider validationProvider;
+  private final OperateClientAdapter operateClientAdapter;
+  private final DynamicPropertyManager dynamicPropertyManager;
+
+  public DefaultInboundConnectorContextFactory(
+      final ObjectMapper mapper,
+      final InboundCorrelationHandler correlationHandler,
+      final SecretProviderAggregator secretProviderAggregator,
+      final ValidationProvider validationProvider,
+      final OperateClientAdapter operateClientAdapter,
+      final DynamicPropertyManager dynamicPropertyManager) {
+    this.objectMapper = mapper;
+    this.correlationHandler = correlationHandler;
+    this.secretProviderAggregator = secretProviderAggregator;
+    this.validationProvider = validationProvider;
+    this.operateClientAdapter = operateClientAdapter;
+    this.dynamicPropertyManager = dynamicPropertyManager;
+  }
+
+  @Override
+  public <T extends InboundConnectorExecutable<?>> InboundConnectorContext createContext(
+      final InboundConnectorDefinitionImpl newConnector,
+      final Consumer<Throwable> cancellationCallback,
+      final Class<T> executableClass) {
+
+    InboundConnectorContext inboundContext =
+        new InboundConnectorContextImpl(
+            secretProviderAggregator,
+            validationProvider,
+            newConnector,
+            correlationHandler,
+            cancellationCallback,
+            objectMapper);
+
+    // Inspects the generic type of InboundConnectorExecutable to determine the type of context
+    // needed and initializes it accordingly.
+    Type[] genericInterfaces = executableClass.getGenericInterfaces();
+    for (Type genericInterface : genericInterfaces) {
+      if (genericInterface instanceof ParameterizedType parameterizedType) {
+        Type[] actualTypeArguments = parameterizedType.getActualTypeArguments();
+        if (actualTypeArguments.length > 0
+            && actualTypeArguments[0].equals(InboundIntermediateConnectorContext.class)) {
+          inboundContext =
+              new InboundIntermediateConnectorContextImpl(
+                  inboundContext, operateClientAdapter, dynamicPropertyManager, objectMapper);
+          break;
+        }
+      }
+    }
```

Comment:
Can we get this covered by unit tests? It feels it can be simplified, but without test I cannot experiment.

Reply:
Please refactor the implementation. You always create an instance of `InboundConnectorContextImpl` even if you throw it away at a later point. Please extract the lookup into a separate method as only create the right context when needed.